swap/htlc.go

```diff
@@ -515,3 +623,239 @@ func (h *HtlcScriptV2) MaxTimeoutWitnessSize() int {
+
+	// The final locking script for the success path in which the receiver
+	// reveals the preimage.
+	SuccessScript []byte
+
+	// The public key for the keyspend path which bypasses the above two
+	// locking scripts.
+	InternalPubKey *btcec.PublicKey
+
+	// The taproot public key which is created with the above 3 inputs.
+	TaprootKey *btcec.PublicKey
+}
+
+// newHTLCScriptV3 constructs a HtlcScipt with the HTLC V3 taproot script.
+func newHTLCScriptV3(cltvExpiry int32, senderHtlcKey,
+	receiverHtlcKey [33]byte, sharedKey *btcec.PublicKey,
+	swapHash lntypes.Hash) (*HtlcScriptV3, error) {
+
+	receiverPubKey, err := btcec.ParsePubKey(
+		receiverHtlcKey[:],
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	senderPubKey, err := btcec.ParsePubKey(
+		senderHtlcKey[:],
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var schnorrSenderKey, schnorrReceiverKey [32]byte
+	copy(schnorrSenderKey[:], schnorr.SerializePubKey(senderPubKey))
+	copy(schnorrReceiverKey[:], schnorr.SerializePubKey(receiverPubKey))
+
+	// Create our success path script, we'll use this separately
+	// to generate the success path leaf.
+	successPathScript, err := GenSuccessPathScript(
+		schnorrReceiverKey, swapHash,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create our timeout path leaf, we'll use this separately
+	// to generate the timeout path leaf.
+	timeoutPathScript, err := GenTimeoutPathScript(
+		schnorrSenderKey, int64(cltvExpiry),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Assemble our taproot script tree from our leaves.
+	tree := txscript.AssembleTaprootScriptTree(
+		txscript.NewBaseTapLeaf(successPathScript),
+		txscript.NewBaseTapLeaf(timeoutPathScript),
+	)
+
+	rootHash := tree.RootNode.TapHash()
+
+	// Calculate top level taproot key.
+	taprootKey := txscript.ComputeTaprootOutputKey(
+		sharedKey, rootHash[:],
+	)
+
+	return &HtlcScriptV3{
+		TimeoutScript:  timeoutPathScript,
+		SuccessScript:  successPathScript,
+		InternalPubKey: sharedKey,
+		TaprootKey:     taprootKey,
+	}, nil
+}
+
+// GenTimeoutPathScript constructs an HtlcScript for the timeout payment path.
+// Largest possible bytesize of the script is 32 + 1 + 2 + 1 = 36.
+//
+//	<senderHtlcKey> OP_CHECKSIGVERIFY <cltvExpiry> OP_CHECKLOCKTIMEVERIFY
+func GenTimeoutPathScript(
+	senderHtlcKey [32]byte, cltvExpiry int64) ([]byte, error) {
+
+	builder := txscript.NewScriptBuilder()
+	builder.AddData(senderHtlcKey[:])
+	builder.AddOp(txscript.OP_CHECKSIGVERIFY)
+	builder.AddInt64(cltvExpiry)
+	builder.AddOp(txscript.OP_CHECKLOCKTIMEVERIFY)
+	return builder.Script()
+}
+
+// GenSuccessPathScript constructs an HtlcScript for the success payment path.
+// Largest possible bytesize of the script is 32 + 5*1 + 20 + 3*1 = 60.
+//
+//	<receiverHtlcKey> OP_CHECKSIGVERIFY
+//	OP_SIZE 32 OP_EQUALVERIFY
+//	OP_HASH160 <ripemd160h(swapHash)> OP_EQUALVERIFY
+//	1 OP_CHECKSEQUENCEVERIFY
+func GenSuccessPathScript(
+	receiverHtlcKey [32]byte, swapHash lntypes.Hash) ([]byte, error) {
```

Comment:
nit: `receiverHtlcKey` should fit on line before, this "break before all params" looks weird to me.
Also, needs empty line after multi-line function declaration.